src/ZenohUTransport.cpp

```diff
@@ -382,9 +382,15 @@ v1::UStatus ZenohUTransport::registerListenerImpl(
 		// We can't use the UUri validators to determine what mode they
 		// are for because a) there is overlap in allowed values between
 		// modes and b) any filter is allowed to have wildcards present.
-		registerResponseListener_(zenoh_key, listener);
 		registerRequestListener_(zenoh_key, listener);
 		registerPublishNotificationListener_(zenoh_key, listener);
+
+		if (sink_filter.has_value()) {
+			// zenoh_key for response listener should be in revert order
+			std::string zenoh_response_key = toZenohKeyString(
+			    getEntityUri().authority_name(), *sink_filter, source_filter);
+			registerResponseListener_(zenoh_response_key, listener);
+		}
```

Comment:
I believe I understand the purpose of this now - it is so that the RPC response listener can be found using the source and sink from the RPC request message's attributes without swapping the position of those URIs. However, I think that it would be better to register the callback to the topic exactly as requested here in `registerListener()`, then in `send()` check the message type attribute and swap the URIs to perform the callback lookup for the query. That will be easier to understand and more directly ties the swap to the actual purpose (response handler lookup for RPC request messages).

Reply:
@gregmedd 
This is not quite true. This key is used for two purposes, to find a response listener in the map and for sending a call to the server (via the `get` method, by this key). But from the server's point of view, we have source and sink inverted.
If we change them here, for the server side they should also be changed (in registerRequestListener_) both here and in rust.
So I would recommend merging this PR as is, and if we need to change the server registration, create a separate issue and change it simultaneously for rust and cpp

Reply:
I agree with @sashacmc that we should have the same behavior in Rust and C++. IMHO, there is no big difference between the two methods. Perhaps, we could go with the current way and ensure the RPC works well first.